plenum/test/replica/test_instance_faulty_processor.py

```diff
@@ -0,0 +1,409 @@
+import pytest
+
+from plenum.common.messages.node_messages import BackupInstanceFaulty
+from plenum.common.types import f
+from plenum.server.backup_instance_faulty_processor import BackupInstanceFaultyProcessor
+from plenum.server.quorums import Quorums
+from plenum.server.replica import Replica
+from plenum.server.suspicion_codes import Suspicions
+from plenum.test.primary_selection.test_primary_selector import FakeNode
+
+
+class FakeReplicas:
+    def __init__(self, node, replicas):
+        self._replicas = replicas
+        self._node = node
+        self.remove_replica_calls = []
+        self.add_replica_calls = []
+
+        self.items = lambda: replicas.items()
+        self.keys = lambda: replicas.keys()
+
+    def remove_replica(self, inst_id):
+        self.remove_replica_calls.append(inst_id)
+        self._replicas.pop(inst_id)
+
+    def add_replica(self, inst_id):
+        self.add_replica_calls.append(inst_id)
+        self._replicas.update(inst_id=Replica(node=self._node, instId=inst_id))
+
+    def __getitem__(self, item):
+        return self._replicas[item]
+
+
+@pytest.fixture(scope="function")
+def backup_instance_faulty_processor(tdir, tconf):
+    node = FakeNode(tdir, config=tconf)
+    node.view_change_in_progress = False
+    node.requiredNumberOfInstances = len(node.replicas)
+    node.allNodeNames = ["Node{}".format(i)
+                         for i in range(1, (node.requiredNumberOfInstances - 1) * 3 + 2)]
+    node.totalNodes = len(node.allNodeNames)
+    node.quorums = Quorums(node.totalNodes)
+    node.name = node.allNodeNames[0]
+    node.replicas = FakeReplicas(node, node.replicas)
+    node.backup_instance_faulty_processor = BackupInstanceFaultyProcessor(node)
+    return node.backup_instance_faulty_processor
+
+
+# TESTS FOR ON_BACKUP_DEGRADATION
+
+
+def test_on_backup_degradation_local(looper,
+                                     backup_instance_faulty_processor):
+    '''
+    1. Call on_backup_degradation() with local strategy for backup performance degradation
+    2. Check that set of degraded instances which was send in on_backup_degradation
+    equals with set in remove_replica call.
+    '''
+    node = backup_instance_faulty_processor.node
+    node.config.REPLICAS_REMOVING_WITH_DEGRADATION = "local"
+    degraded_backups = [1, 2]
+
+    backup_instance_faulty_processor.on_backup_degradation(degraded_backups)
+
+    assert not (set(node.replicas.remove_replica_calls) - set(degraded_backups))
```

Comment:
Why can't we just compare two sets here? `assert set(node.replicas.remove_replica_calls) == set(degraded_backups)`